File: Stack/Leetcode/Hard/SlidingWindowMaximum/SlidingWindowMaximum.java
// https://leetcode.com/problems/sliding-window-maximum/

import java.util.ArrayDeque;
import java.util.Deque;

public class SlidingWindowMaximum {

    public int[] maxSlidingWindow(int[] nums, int k) {
        Deque<Integer> dq = new ArrayDeque<>();
        int l=0, r=0;
        int[] ans = new int[nums.length - k + 1];
        while(r < nums.length) {
            while(!dq.isEmpty() && nums[dq.getLast()] <= nums[r]) {
                dq.removeLast();
            }
            dq.offerLast(r);
            if(dq.getFirst() < l) {
                dq.removeFirst();
            }
            if(r - l + 1 == k) {
                ans[l] = nums[dq.getFirst()];
                l++;
            }
            r++;
        }
        return ans;
    }
}

File: Stack/Leetcode/Hard/SlidingWindowMaximum/SlidingWindowMaximum.cpp
// https://leetcode.com/problems/sliding-window-maximum/
#include<bits/stdc++.h>
using namespace std;
vector<int> maxSlidingWindow(vector<int> &nums, int k)
{
    deque<int> dq;
    vector<int> ans;
    int l = 0, r = 0;
    while (r < nums.size())
    {
        while (!dq.empty() && nums[dq.back()] <= nums[r])
        {
            dq.pop_back();
        }
        dq.push_back(r);
        if (dq.front() < l)
        {
            dq.pop_front();
        }
        if (r - l + 1 == k)
        {
            ans.push_back(nums[dq.front()]);
            l++;
        }
        r++;
    }
    return ans;
}